autoscale/src/main/java/com/sequenceiq/periscope/monitor/evaluator/load/YarnResponseUtils.java

```diff
@@ -0,0 +1,71 @@
+package com.sequenceiq.periscope.monitor.evaluator.load;
+
+import static com.sequenceiq.periscope.monitor.evaluator.ScalingConstants.DEFAULT_MAX_SCALE_UP_STEP_SIZE;
+
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+import com.sequenceiq.periscope.model.yarn.YarnScalingServiceV1Response;
+
+@Component
+public class YarnResponseUtils {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(YarnResponseUtils.class);
+
+    public List<String> getYarnRecommendedDecommissionHostsForHostGroup(String clusterCrn, YarnScalingServiceV1Response yarnResponse,
+            Map<String, String> hostFqdnsToInstanceId, int maxAllowedDownScale, Optional<Integer> mandatoryDownScaleCount) {
+        Set<String> consideredNodeIds = new HashSet<>();
+        Integer allowedDownscale = Math.max(maxAllowedDownScale, mandatoryDownScaleCount.orElse(0));
+        List<String> decommissionNodes = yarnResponse.getScaleDownCandidates().orElse(List.of()).stream()
+                .sorted(Comparator.comparingInt(YarnScalingServiceV1Response.DecommissionCandidate::getAmCount))
+                .map(YarnScalingServiceV1Response.DecommissionCandidate::getNodeId)
+                .map(nodeFqdn -> nodeFqdn.split(":")[0])
+                .filter(s -> hostFqdnsToInstanceId.keySet().contains(s))
+                .map(nodeFqdn -> hostFqdnsToInstanceId.get(nodeFqdn))
+                .limit(allowedDownscale)
+                .map(nodeId -> {
+                    consideredNodeIds.add(nodeId);
+                    return nodeId;
+                })
+                .collect(Collectors.toList());
+
+        //ForcedCandidates initialization is required since Yarn currently does not support hostGroup based recommendation.
+        int forcedCandidatesCount = mandatoryDownScaleCount.orElse(0) - decommissionNodes.size();
+        if (forcedCandidatesCount > 0) {
```

Comment:
Above would still not guarantee required number of nodes of a particular hostGroup it just includes more nodes across hostGroups. 
Also passing a high value would still be forced downscale because Yarn returns all nodes to match the requested N irrespective of if they are really decommissionable. 
Once COMPX-3425 is fixed which is expected soon, this condition would not be processed. If there are any yarn downscale recommendation issues after COMPX-3425 we can figure out from below log and also would like to avoid fixing all Yarn Bugs in periscope.

Reply:
Yup, it is the same as sending the entire cluster size. Doesn't YARN do some smart sorting of the list though before sending it back? Think that is a perfectly fair workaround while COMPX-3425 gets implemented.
Why is the same as existing behaviour of Schedule Based downscaling?
- If there are real nodes which can be downscaled, those will be at the top of the list.
- The rest will hopefully be ordered by AM count / amount of work.
On the Periscope side, filtering by hostGroup would be adequate.
That's how I was expecting this to be used anyway.

I buy the argument of waiting till COMPX-3425 though; we'll have to be careful about compatibility, or go and say that AutoScaling can only be used with 7.2.2 (and provide Experian with hotfixes)

Reply:
Ok, based on our discussion and input from yarn team, it makes sense to pass N = cluster size which helps us to choose better candidates to downscale, hence updated code to pass the same.